File: web/src/drift-corrector.js
const INSYNC_THRESHOLD_MS = 50;
const SOFT_SEEK_THRESHOLD_MS = 1500;
const PROPORTIONAL_GAIN = 1 / 5000;
const MIN_SPEED = 0.85;
const MAX_SPEED = 1.15;
const SMOOTH_FACTOR = 0.5;

let smoothedDrift = 0;

export function evaluate(currentPosition, expectedPosition) {
  const rawDrift = currentPosition - expectedPosition;
  const magnitude = Math.abs(rawDrift);

  if (magnitude < INSYNC_THRESHOLD_MS) {
    smoothedDrift = 0;
    return { type: 'inSync', driftMs: rawDrift };
  }

  smoothedDrift = SMOOTH_FACTOR * rawDrift + (1 - SMOOTH_FACTOR) * smoothedDrift;
  const drift = Math.round(smoothedDrift);
  const mag = Math.abs(drift);

  if (mag < SOFT_SEEK_THRESHOLD_MS) {
    const speedAdjust = Math.min(Math.max(-drift * PROPORTIONAL_GAIN, MIN_SPEED - 1), MAX_SPEED - 1);
    const speed = Math.min(Math.max(1 + speedAdjust, MIN_SPEED), MAX_SPEED);
    return { type: 'softCorrect', driftMs: drift, speed };
  } else {
    return { type: 'softSeek', driftMs: drift, expectedPosition };
  }
}

export function applyDriftCorrection(video, expectedPositionMs, isPaused, onStatus) {
  const currentPosition = video.currentTime * 1000;
  const action = evaluate(currentPosition, expectedPositionMs);

  if (isPaused) return;

  switch (action.type) {
    case 'inSync':
      video.playbackRate = 1.0;
      if (onStatus) onStatus('synced');
      break;

    case 'softCorrect':
      video.playbackRate = action.speed;
      if (onStatus) onStatus('correcting');
      break;

    case 'softSeek':
      const currentMs = video.currentTime * 1000;
      const targetMs = currentMs + (action.expectedPosition - currentMs) * 0.8;
      video.currentTime = Math.max(targetMs / 1000, 0);
      video.playbackRate = 1.0;
      if (onStatus) onStatus('correcting');
      break;
  }
}

export function resetDriftOnCommand() {
  smoothedDrift = 0;
}

export function cleanup() {
  smoothedDrift = 0;
}

File: app/src/main/java/com/nityam/movsync/data/sync/DriftCorrector.kt
package com.nityam.movsync.data.sync

import androidx.media3.common.Player
import kotlin.math.abs

class DriftCorrector {
    companion object {
        private const val INSYNC_THRESHOLD_MS = 50L
        private const val SOFT_SEEK_THRESHOLD_MS = 800L
        private const val PROPORTIONAL_GAIN = 1f / 5000f
        private const val MIN_SPEED = 0.85f
        private const val MAX_SPEED = 1.15f
        private const val SMOOTH_FACTOR = 0.5
    }

    private var smoothedDrift = 0.0

    fun evaluate(currentPositionMs: Long, expectedPositionMs: Long): DriftAction {
        val rawDrift = currentPositionMs - expectedPositionMs
        val magnitude = abs(rawDrift)

        if (magnitude < INSYNC_THRESHOLD_MS) {
            smoothedDrift = 0.0
            return DriftAction.InSync(rawDrift)
        }

        smoothedDrift = SMOOTH_FACTOR * rawDrift + (1.0 - SMOOTH_FACTOR) * smoothedDrift
        val drift = smoothedDrift.toLong()
        val mag = abs(drift)

        if (mag < SOFT_SEEK_THRESHOLD_MS) {
            val speedAdjust = (-drift * PROPORTIONAL_GAIN).coerceIn(
                MIN_SPEED - 1f, MAX_SPEED - 1f
            )
            val speed = (1f + speedAdjust).coerceIn(MIN_SPEED, MAX_SPEED)
            return DriftAction.SoftCorrect(drift, speed)
        } else {
            return DriftAction.SoftSeek(drift, expectedPositionMs)
        }
    }

    fun apply(
        player: Player,
        expectedPositionMs: Long,
        isPaused: Boolean,
        onStatus: (SyncStatus) -> Unit
    ) {
        if (isPaused) return

        when (val action = evaluate(player.currentPosition, expectedPositionMs)) {
            is DriftAction.InSync -> {
                player.setPlaybackSpeed(1f)
                onStatus(SyncStatus.Synced)
            }
            is DriftAction.SoftCorrect -> {
                player.setPlaybackSpeed(action.speed)
                onStatus(SyncStatus.Correcting)
            }
            is DriftAction.SoftSeek -> {
                val currentPos = player.currentPosition
                val target = currentPos + ((action.expectedPosition - currentPos) * 1.0f).toLong()
                player.seekTo(target)
                player.setPlaybackSpeed(1f)
                onStatus(SyncStatus.Correcting)
            }
        }
    }

    fun resetOnCommand() {
        smoothedDrift = 0.0
    }
}

sealed interface DriftAction {
    data class InSync(val driftMs: Long) : DriftAction
    data class SoftCorrect(val driftMs: Long, val speed: Float) : DriftAction
    data class SoftSeek(val driftMs: Long, val expectedPosition: Long) : DriftAction
}

enum class SyncStatus {
    Synced,
    Correcting,
    Reconnecting
}
